File: algorithm_practice_python/leetcode/arrays101/2367.py
from typing import List


class Solution:
    def arithmeticTriplets(self, nums: List[int], diff: int) -> int:
        """
            arithmetic triplets
            1. 0 <= i < j < k < len(nums)
            2. nums[j]-nums[i] = diff
            3. nums[k]-nums[j] = diff
        """
        ans = 0
        length = len(nums)
        for i in range(length):
            if nums[i] + diff in nums and nums[i] + 2 * diff in nums:
                ans += 1
        return ans

    def sol2(self, nums, diff):
        seen = set(nums)
        return sum(num - diff in seen and num - diff * 2 in seen for num in seen)

if __name__ == '__main__':
    nums = [0, 1, 4, 6, 7, 10]
    diff = 3
    ans = Solution().sol2(nums, diff)
    print(f'{ans}, 2')

File: algorithm_practice_typescript/2367.ts
function arithmeticTriplets(nums: number[], diff: number): number {
    let ans = 0
    for (let i = 0; i < nums.length; i++) {
        if (nums.includes(nums[i] + diff) && nums.includes(nums[i] + diff * 2)) {
            ans++
        }
    }
    return ans
}

function sol2(nums: number[], diff: number): number {
    let ans = 0
    let hash = new Set()

    for (let i = 0; i < nums.length; i++) {
        const currentNum = nums[i]

        if (hash.has(currentNum - diff) && hash.has(currentNum - diff * 2)) {
            ans++
        }
        hash.add(currentNum)
    }

    return ans
}

const n: number[] = [0,1,4,6,7,10]
const diff = 3

console.log(sol2(n, diff))
